src/main/java/com/uid2/admin/vertx/service/ClientSideKeypairService.java

```diff
@@ -0,0 +1,250 @@
+
+import java.security.KeyPair;
+import java.time.Instant;
+import java.util.*;
+import java.util.stream.Collectors;
+
+public class ClientSideKeypairService implements IService, IKeypairManager {
+    private final AuthMiddleware auth;
+    private final Clock clock;
+
+    private final WriteLock writeLock;
+    private final ClientSideKeypairStoreWriter storeWriter;
+    private final RotatingClientSideKeypairStore keypairStore;
+    private final RotatingSiteStore siteProvider;
+    private final IKeypairGenerator keypairGenerator;
+    private static final Logger LOGGER = LoggerFactory.getLogger(ClientSideKeypairService.class);
+
+    public ClientSideKeypairService(AuthMiddleware auth,
+                                    WriteLock writeLock,
+                                    ClientSideKeypairStoreWriter storeWriter,
+                                    RotatingClientSideKeypairStore keypairStore,
+                                    RotatingSiteStore siteProvider,
+                                    IKeypairGenerator keypairGenerator,
+                                    Clock clock) {
+        this.auth = auth;
+        this.writeLock = writeLock;
+        this.storeWriter = storeWriter;
+        this.keypairStore = keypairStore;
+        this.keypairGenerator = keypairGenerator;
+        this.siteProvider = siteProvider;
+        this.clock = clock;
+    }
+
+    @Override
+    public void setupRoutes(Router router) {
+        router.post("/api/client_side_keypairs/add").blockingHandler(auth.handle((ctx) -> {
+            synchronized (writeLock) {
+                this.handleAddKeypair(ctx);
+            }
+        }, Role.ADMINISTRATOR));
+        router.post("/api/client_side_keypairs/update").blockingHandler(auth.handle((ctx) -> {
+            synchronized (writeLock) {
+                this.handleUpdateKeypair(ctx);
+            }
+        }, Role.ADMINISTRATOR));
+        router.get("/api/client_side_keypairs/list").handler(
+                auth.handle(this::handleListAllKeypairs, Role.ADMINISTRATOR));
+        router.get("/api/client_side_keypairs/:subscriptionId").handler(
+                auth.handle(this::handleListKeypair, Role.ADMINISTRATOR)
+        );
+    }
+
+    private void handleAddKeypair(RoutingContext rc){
+        final JsonObject body = rc.body().asJsonObject();
+        final Integer siteId = body.getInteger("site_id");
+        final String contact = body.getString("contact");
+        final boolean disabled = body.getBoolean("disabled", false);
+        if(contact == null || siteId == null) {
+            ResponseUtil.error(rc, 400, "Required parameters: site_id, contact");
+            return;
+        }
+        if(siteProvider.getSite(siteId) == null) {
+            ResponseUtil.error(rc, 404, "site_id: " + siteId + " not valid");
+            return;
+        }
+        String normalizedContactEmail = InputUtil.normalizeEmailString(contact);
+        if(normalizedContactEmail == null){
```

Comment:
why do we need to normalize the email?

Reply:
I suppose it isn't necessary, but it can provide some amount of typo protection in the event the email can't be normalized. If normalization succeed then it may just look nicer.